presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolAllocator.java

```diff
@@ -126,9 +126,9 @@ public Symbol newSymbol(Field field)
         return newSymbol(nameHint, field.getType());
     }
 
-    public Map<Symbol, Type> getTypes()
+    public TypeProvider getTypes()
     {
-        return symbols;
+        return TypeProvider.mutable(symbols);
```

Comment:
why is this `mutable`? I haven't seen usage of `getAllTypes` which modify underlying table, what have I missed?

Reply:
When we add a new symbol to `SymbolAllocator` it adds a new entry to `symbols` and it has to be reflected in the TypeProvider so it should be mutable. This should reduce the deep copy of 'Map<Symbol, Type>` which happens at `ExpressionAnalyzer` (It takes more time when we have a more number of wide tables in a given query thereby increasing the planning time).

Reply:
@kokosing, the TypeProvider itself is not mutable, but the underlying map is. That was the best name I could come up with at the time, but I think I'll rename it to "viewOf", "asView" or something like that.